### PR TITLE
bpo-34616: add flags to allow top-level-await 

### PR DESCRIPTION
**Pull request by Carreau**
Based on a work with @1st1 during PyCon 2019; and can be seen used in https://github.com/ipython/ipython/pull/11713 (which would dramatically reduce thee complexity and shave an extra few hundreds lines of IPython if we were supporting 3.8+ only).

https://bugs.python.org/issue34616

### CODE REVIEW

**Comment by Carreau** [comment]
If You wish to try it, here is a fake-repl session script that work as you would expect; 

```
import asyncio
from ast import PyCF_ALLOW_TOP_LEVEL_AWAIT
from inspect import CO_COROUTINE
from textwrap import dedent

is_async = (lambda co: co.co_flags & CO_COROUTINE == CO_COROUTINE)
g = {}
l = {}

@asyncio.coroutine
def f():
    for inp in ['import asyncio',
                'print("before")',
                dedent('''
                print("multi statement in eval...")
                res = await asyncio.sleep(0, result="slept")
                print(res)
                print(res[::-1])
                '''),
                'print("finished")']:
        co = compile(inp, '', 'exec', flags=PyCF_ALLOW_TOP_LEVEL_AWAIT)
        if is_async(co):
            yield from eval(co, l, g)
        else:
            exec(co, l,g)

import asyncio
asyncio.run(f())
```

The weird-ish think is the need to use `yield from eval(...)` when running async-code.. but close enough; 

Works as well with trio (cc @njsmith) if you use 
```
def _trio_runner(async_fn):
    import trio

    async def loc(coro):
        """
        We need the dummy no-op async def to protect from
        trio's internal. See https://github.com/python-trio/trio/issues/89
        """
        return await coro

    return trio.run(loc, async_fn)

_trio_runner(f())
```

**Comment by njsmith** [comment]
CC @oremanj

File: asyncex [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import asyncio
```
Comment (1st1) [inline asyncex]: This file shouldn't be part of the PR.  We'll need to add a couple of unit tests though.

File: trycomp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Comment (1st1) [inline trycomp]: This file shouldn't be part of the PR.

File: Python/compile.c [hunk 3]
```diff
@@ -4759,12 +4765,20 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
...
+                return compiler_error(c, "'await' outside function");
+            }
 
-        if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION &&
-                c->u->u_scope_type != COMPILER_SCOPE_COMPREHENSION)
-            return compiler_error(c, "'await' outside async function");
+            c->u->u_ste->ste_coroutine = 1;
+        }
+
+        if (!(c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT)) {
```
Comment (1st1) [inline Python/compile.c]: This check should be the top-level if; then nest the other two checks inside.

File: Python/compile.c [hunk 4]
```diff
@@ -4759,12 +4765,20 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
...
-            return compiler_error(c, "'await' outside function");
+        if (c->u->u_ste->ste_type != FunctionBlock) {
+            if (!(c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT)) {
+                return compiler_error(c, "'await' outside function");
+            }
 
-        if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION &&
-                c->u->u_scope_type != COMPILER_SCOPE_COMPREHENSION)
-            return compiler_error(c, "'await' outside async function");
+            c->u->u_ste->ste_coroutine = 1;
```
Comment (1st1) [inline Python/compile.c]: I'm not sure if this is actually necessary, try to remove this line and see if the tests pass (and add tests :)).

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Python/compile.c [hunk 5]
```diff
@@ -4759,12 +4763,17 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
...
 
-        if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION &&
-                c->u->u_scope_type != COMPILER_SCOPE_COMPREHENSION)
-            return compiler_error(c, "'await' outside async function");
+            if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION && 
+                    c->u->u_scope_type != COMPILER_SCOPE_COMPREHENSION){
+                return compiler_error(c, "'await' outside async function");
+            }
+        }
+        c->u->u_ste->ste_coroutine = 1;
```
Comment (1st1) [inline Python/compile.c]: I'm still not sure if this line is needed, but if it is, it should only be set when `(c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT)` is true.

File: Python/compile.c [hunk 6]
```diff
@@ -5699,6 +5708,10 @@ compute_code_flags(struct compiler *c)
     /* (Only) inherit compilerflags in PyCF_MASK */
     flags |= (c->c_flags->cf_flags & PyCF_MASK);
 
+    if ((c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT) && c->u->u_ste->ste_coroutine) {
+        flags |= CO_COROUTINE;
```
Comment (Carreau) [inline Python/compile.c:5726]: This thing has some weird effect and turn async_generators into coroutine;  I'm going to change that to `!ste->ste_generator`

File: Python/compile.c (hunk 4 above)
Comment (Carreau) [reply Python/compile.c]: ... well it actually is or I get 
```
unhandled exception during asyncio.run() shutdown
task: <Task finished name='Task-2' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError("can't send non-None value to a just-started coroutine")>
RuntimeError: can't send non-None value to a just-started coroutine
```

Not quite sure I understand exactly why...

**Comment by Carreau** [comment]
You were right that the ste->ste_coroutine was not necessary in AwaitKind, though it is necessary in `async_with` and `async_for`. I've removed some of the unnecessary files, and shuffled the `if`s to be a bit more readable.


Here is a short asyncio-repl with some example inputs: 
https://gist.github.com/Carreau/9d8626ab0cf356b63d3481f28a394562

**Comment by Carreau** [comment]
I have made the requested changes; please review again

**Comment by Carreau** [comment]
I've pushed again to restart the failing test on travis; all seem to work. Let me know what else I can do; I'd love to get that in 3.8.

File: Doc/library/functions.rst [hunk 7]
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
    can be found as the :attr:`~__future__._Feature.compiler_flag` attribute on
    the :class:`~__future__._Feature` instance in the :mod:`__future__` module.
 
+   The optional argument *flags* also control wether the compiled source is
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
+   object has ``CO_COROUTINE`` set in ``co_code``, and can be interacitvely
```
Comment (tirkarthi) [inline Doc/library/functions.rst]:
```suggestion
   object has ``CO_COROUTINE`` set in ``co_code``, and can be interactively
```

File: Doc/library/functions.rst [hunk 8]
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
    can be found as the :attr:`~__future__._Feature.compiler_flag` attribute on
    the :class:`~__future__._Feature` instance in the :mod:`__future__` module.
 
+   The optional argument *flags* also control wether the compiled source is
```
Comment (tirkarthi) [inline Doc/library/functions.rst]:
```suggestion
   The optional argument *flags* also controls whether the compiled source is
```

File: Lib/test/test_builtin.py [hunk 9]
```diff
@@ -358,6 +361,32 @@ def f(): """doc"""
                 rv = ns['f']()
                 self.assertEqual(rv, tuple(expected))
 
+    def test_compile_top_level_await(self):
+        for mode in ('single', 'exec'):
+            for num,code_sample in enumerate(['''await sleep(0)''',
```
Comment (tirkarthi) [inline Lib/test/test_builtin.py]: Minor nit : num is not used in test so `enumerate` seems to be not needed . Perhaps the list used in enumerate can be defined as a variable outside of for loop and just iterated?

File: Lib/test/test_builtin.py [hunk 10]
```diff
@@ -358,6 +361,32 @@ def f(): """doc"""
...
+                                '''async for i in range(10):
+                                       print(i)''',
+                                '''async with asyncio.Lock() as l:
+                                       pass''']):
+                source = dedent(code_sample)
+                with self.assertRaises(SyntaxError,
+                                  msg='source={!r} mode={!r})'.format(source, mode)):
```
Comment (tirkarthi) [inline Lib/test/test_builtin.py]: Align msg with `SyntaxError` ? PEP 8 enforces 80 character limit per line but I will leave it to reviewer on the same.

File: Lib/test/test_builtin.py (hunk 9 above)
Comment (Carreau) [reply Lib/test/test_builtin.py]: Good point, I used it to debug which sample(s) were wrong and did not removed it, will do.

File: Lib/test/test_builtin.py (hunk 10 above)
Comment (Carreau) [reply Lib/test/test_builtin.py]: I've done that and used `itertools.product` to remove one level of indent. I'm still slightly over.

Comment (tirkarthi) [reply Lib/test/test_builtin.py]: Perhaps have `msg` as a variable outside of context manager.

```python
msg = 'source={!r} mode={!r})'.format(source, mode)

with self.assertRaises(SyntaxError, msg):
    compile(source, '?' , mode)
```

or even better master has f-string debugging merged into master where `f"{source=}"` expands into `f"source={source!r}"`. So you can rebase onto master and write below : 

```python
with self.assertRaises(SyntaxError, msg=f'{source=} {mode=})'): # 75 characters
    compile(source, '?' , mode)
```

Comment (Carreau) [reply Lib/test/test_builtin.py]: Yep good idea. Done that and rebased on master.

File: Doc/library/functions.rst [hunk 11]
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
    can be found as the :attr:`~__future__._Feature.compiler_flag` attribute on
    the :class:`~__future__._Feature` instance in the :mod:`__future__` module.
 
+   The optional argument *flags* also controls whether the compiled source is
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
+   object has ``CO_COROUTINE`` set in ``co_code``, and can be interactively
+   executed via ``yield from eval(code_object)``.
```
Comment (1st1) [inline Doc/library/functions.rst]:
```suggestion
   executed via ``await eval(code_object)``.
```

File: Doc/library/functions.rst [hunk 12]
```diff
@@ -290,6 +296,10 @@ are always available.  They are listed here in alphabetical order.
       Previously, :exc:`TypeError` was raised when null bytes were encountered
       in *source*.
 
+   .. versionadded:: 3.8
+      *flags* can now be used to accept source that would contain a top-level
```
Comment (1st1) [inline Doc/library/functions.rst]: ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` can now be passed in *flags* to enable support
for top-level await, async for, and async with.

File: Lib/test/test_builtin.py [hunk 13]
```diff
@@ -358,6 +362,43 @@ def f(): """doc"""
...
+
+    def test_compile_async_generator(self):
+        """
+        With the PyCF_ALLOW_TOP_LEVEL_AWAIT flag added in 3.8, we want to
+        make sure AsyncGenerators are still properly not marked with CO_COROUTINE
+        """
+        co = compile(dedent("""async def ticker():
+                for i in range(10):
+                    yield i
+                    await asyncio.sleep(0)"""), '?', 'exec', flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT)
```
Comment (1st1) [inline Lib/test/test_builtin.py]: Please make sure all new lines fit the 79 characters limit.

File: Python/compile.c (hunk 6 above)
Comment (1st1) [reply Python/compile.c:5726]: Please fix to fit 79 characters

File: Python/compile.c [hunk 14]
```diff
@@ -4564,7 +4566,9 @@ compiler_async_with(struct compiler *c, stmt_ty s, int pos)
     withitem_ty item = asdl_seq_GET(s->v.AsyncWith.items, pos);
 
     assert(s->kind == AsyncWith_kind);
-    if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION) {
+    if (c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT){
+        c->u->u_ste->ste_coroutine = 1;
```
Comment (1st1) [inline Python/compile.c:4570]: Have you checked if this `c->u->u_ste->ste_coroutine = 1;` line is needed?  Does it work without it?

File: Lib/test/test_builtin.py [hunk 15]
```diff
@@ -358,6 +362,43 @@ def f(): """doc"""
...
+               pass''']
+        for mode, code_sample in product(modes,code_samples):
+            source = dedent(code_sample)
+            with self.assertRaises(SyntaxError, msg=f"{source=} {mode=}"):
+                compile(source, '?' , mode)
+
+            co = compile(source, '?', mode, flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT)
+
+            self.assertEqual(co.co_flags & CO_COROUTINE, CO_COROUTINE,
+                             msg=f"{source=} {mode=}")
```
Comment (1st1) [inline Lib/test/test_builtin.py]: Please add a functional test, that wraps the created code object in a FunctionType and then manually iterates the coroutine (or with asyncio) and checks the result.

File: Python/compile.c (hunk 14 above)
Comment (Carreau) [reply Python/compile.c:4570]: yes it is necessary, otherwise: 

```
>>> async with acm() as delta:
>>>     count += delta

Traceback (most recent call last):
  File "minirepl.py", line 97, in f
    co = compile(inp, '', 'single', flags=PyCF_ALLOW_TOP_LEVEL_AWAIT)
  File "<string>", line 2
SyntaxError: 'async with' outside async function
```

Comment (1st1) [reply Python/compile.c:4570]: OK, cool, let's keep it then :)

Comment (Carreau) [reply Python/compile.c:4570]: And there is be a test in `Lib/test/test_builtin.py` for that  in case you ask.

File: Python/compile.c (hunk 6 above)
Comment (Carreau) [reply Python/compile.c:5726]: Done.

File: Lib/test/test_builtin.py [hunk 16]
```diff
@@ -358,6 +364,72 @@ def f(): """doc"""
...
+                globals_ = {'asyncio': asyncio, 'a':0, 'arange': arange}
+                async_f = FunctionType(co, globals_)
+                asyncio.run(async_f())
+                self.assertEqual(globals_['a'], 1)
+
+                # test we can await-eval,
+                globals_ = {'asyncio': asyncio, 'a':0, 'arange': arange}
+                asyncio.run(eval(co, globals_))
+                self.assertEqual(globals_['a'], 1)
+        except Exception:
```
Comment (Carreau) [inline Lib/test/test_builtin.py]: I rewrote most of this, to also run via a `FunctionType` – though I believe creating FunctionTypes directly is not supposed to be done by usercode. and via `run(eval())`

File: Doc/library/functions.rst (hunk 12 above)
Comment (Carreau) [reply Doc/library/functions.rst]: Done.

File: Lib/test/test_builtin.py (hunk 13 above)
Comment (Carreau) [reply Lib/test/test_builtin.py]: Done.

File: Lib/test/test_builtin.py (hunk 15 above)
Comment (Carreau) [reply Lib/test/test_builtin.py]: I've added 2 with `FunctionTypes` and `eval()`

File: Doc/library/functions.rst [hunk 17]
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
    can be found as the :attr:`~__future__._Feature.compiler_flag` attribute on
    the :class:`~__future__._Feature` instance in the :mod:`__future__` module.
 
+   The optional argument *flags* also controls whether the compiled source is
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
```
Comment (Carreau) [inline Doc/library/functions.rst]: Note to self. Double "the" here.

File: Doc/library/functions.rst [hunk 18]
```diff
@@ -290,6 +296,10 @@ are always available.  They are listed here in alphabetical order.
       Previously, :exc:`TypeError` was raised when null bytes were encountered
       in *source*.
 
+   .. versionadded:: 3.8
+      ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` can now be passed in flags to enable
+      support for top-level await, async for, and async with.
```
Comment (Carreau) [inline Doc/library/functions.rst]: note to self. Put `async ...` between backquotes.

**Comment by Carreau** [comment]
All is up to date, and test are passing. Ready for one next iteration of the review.

**Comment by 1st1** [comment]
> All is up to date, and test are passing. Ready for one next iteration of the review.

I think there's no need for that! Thank so much for working on this and pushing this through, Matthias.  Would be great to see more contributions from you!  We need help with asyncio and many other CPython modules.

**Comment by 1st1** [comment]
@Carreau looks like `python -m test.test_tools` fails with this merged.  Do you have time to take a look?

**Comment by ned-deily** [comment]
Please double check the bpo issue number. This PR appears to be unrelated.

**Comment by 1st1** [comment]
> Please double check the bpo issue number. This PR appears to be unrelated.

@ned-deily It's related -- the bpo is about adding an async exec and this PR is the way to do that.  I'm working on unbreaking buildbots in https://github.com/python/cpython/pull/13473

**Comment by ned-deily** [comment]
Sorry, I should have been more specific.  For some reason, PR #13148 is also showing up as a PR under https://bugs.python.org/issue25234.  I'm not yet sure why.

**Comment by Carreau** [comment]
Yes I can have a look... I'll try to catch up on what happen in the last few hours.

**Comment by ned-deily** [comment]
(It looks like the spurious bpo entry was caused by something in one of the intermediate commits. I'll just delete the PR entry from bpo-25234.)

**Comment by 1st1** [comment]
@Carreau 

> Yes I can have a look... I'll try to catch up on what happen in the last few hours.

Never mind, I think I've fixed it already.

**Comment by Carreau** [comment]
> Never mind, I think I've fixed it already.

Ok, Thanks  apologies for the breakage...

**Comment by 1st1** [comment]
> Ok, Thanks apologies for the breakage...

NP, it's not your fault at all.  Strangely, `test_tools` isn't run by our main CI, therefore the "unexpected" breakage.

**Comment by Carreau** [comment]
> I think there's no need for that! Thank you so much for working on this and pushing this through, Matthias. 

Thanks for guiding me, it was way easier with an expert on my side. And looking forward to contribute more.




Would be great to see more contributions from you! We need help with asyncio and many other CPython modules.

> NP, it's not your fault at all. Strangely, `test_tools` isn't run by our main CI, therefore the "unexpected" breakage.

Is there an issue for that, and the fact the f-debug is not supported by test_unparse ?

**Comment by 1st1** [comment]
> Thanks for guiding me, it was way easier with an expert on my side. And looking forward to contribute more.

Great! 🙏

> Is there an issue for that, and the fact the f-debug is not supported by test_unparse ?

I'm not sure, but would appreciate if you could double check that.

**Comment by gvanrossum** [comment]
This is great, but I wish there was also a documented way to generate an AST for such code. Currently the feature of letting `compile()` generate an AST is undocumented, and the documented way to do this is `ast.parse()`. Currently the latter has no way to pass arbitrary flags, so we would have to resort to undocumented(*) behavior if we wanted an AST for code with toplevel `await`. Could you add a flag to `ast.parse()` that also passes this flag? (Alternatively, rewrite the docs for `compile()` to document `ast.PyCF_ONLY_AST` and `ast.PyCF_TYPE_COMMENTS`.)

(*) Well, it's not completely undocumented -- the ast module's docs reference this. But `compile()`'s docs don't.

**Comment by Carreau** [comment]
I believe that since 3.7, the SyntaxError is raised only by `compile()` when requesting a code-object, `ast.parse()` should behave fine with top-level await without any specific flags: 

```
$ python
Python 3.7.1 (default, Dec 14 2018, 13:28:58)
[Clang 4.0.1 (tags/RELEASE_401/final)] :: Anaconda custom (64-bit) on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ast
>>>mod =  ast.parse('await sleep(0)')
>>> mod.body[0].value
<_ast.Await object at 0x107df50b8>
```

Do you want clarification in the docs; or still add a flag to ast-parse to prevent top-level-await unless opted-in ?

**Comment by gvanrossum** [comment]
Huh, I didn't realize that some of these "syntax" errors come from
compile.c (not just that one, also about e.g. return outside function).

Given this, I don't think it makes sense to tweak ast.parse to *prevent*
this. I am dithering about whether to document all the flags supported by
compile() with the compile() function itself -- IIRC they are intentionally
not documented there currently because we don't want to place the burden of
supporting all of this on all other Python implementations.

So now I'm not sure whether there's any action item at all. What do you
think? What do others think?

**Comment by Carreau** [comment]
> Huh, I didn't realize that some of these "syntax" errors come from compile.c (not just that one, also about e.g. return outside function).

To be fair, some of these used to happen both in making the ast and compile.c.

I'm inclined toward leaving this as-is for 3.8, and potentially refine for 3.9. It is a pretty narrow use case; and seeing what people are using this for we can refine the API(s)/documentation. 

This is still [quite tricky to use if you look at Yuri Asyncio REPL.](https://github.com/python/cpython/pull/13472/files#diff-b857c21ce7ca59e68e77b8624808d349R50), so maybe once this stabilise we can actually recommend a specific way of using it.

**Comment by gvanrossum** [comment]
OK, let's leave it alone.
